content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md

```diff
@@ -0,0 +1,804 @@
+
+With JIT linking, at run time, symbolic references can be resolved to existing symbols (from the newly JIT’d code), or to newly JIT'd symbols (from the pre-compiled code). The below toy example shows what this looks like in code.
+
+```c
+// Let's assume we have the following, rather contrived,
+// C++ program that wants to add 2 numbers, but wants to use
+// an `add` function from a relocatable object file supplied by
+// the user.
+//
+// Let's also assume that the add function in the user-supplied
+// relocatable object will reference a symbol named `MAGIC` in its
+// definition.
+
+const int MAGIC = 42;
+
+int main(int argc, char* argv[]) {
+    int a = 1;
+    int b = 2;
+    
+    // Read the path of the user supplied relocatable object.
+    string userSuppliedObjectPath = ...;
+    
+    // Initialize your JIT class that uses JIT linking under the hood.
+    JIT J;
+    
+    // Add the relocatable object to your JIT.
+    J.addObject(userSuppliedObjectPath);
+    
+    // Lookup the `add` function in the newly added JIT object.
+    // Once all symbolic references within the user supplied object
+    // are resolved, the content is fixed up and emitted to memory.
+    // And we can then get a pointer to the `add` function.
+    auto *add = (int(*)(int, int))J.lookup("add").getAddress();
+    
+    // At this point the symbolic reference to `MAGIC` in add's
+    // definition must have been resolved to the memory address
+    // of the constant `MAGIC` that we defined in this program.
+
+    // Run the add function found in the JIT module.
+    int result = add(a, b);
+}
+```
+That said, JIT linking by itself, is not something that is very useful for an end user. JIT linking is an enabler 
+for certain use-cases with pre-compiled languages (some use-cases exist for JIT-compiled languages as well[^2]).
+
+1. **JIT compilers** (think of something like the JIT compilation component of the [Java Hotspot VM](https://en.wikipedia.org/wiki/HotSpot_(virtual_machine)), but for a 
+   statically compiled language)
+2. **Debugger expression evaluators** (such as the LLDB expression evaluator)
+3. **REPLs** (such as [Cling](https://root.cern/cling/) and the currently experimental[^3] [Clang-REPL](https://clang.llvm.org/docs/ClangRepl.html))
+4. **Stand alone scripts** (such as the [Swift scripts](https://jblevins.org/log/swift), where the JIT linking is 
+   used to add an [immediate](https://github.com/apple/swift/blob/main/lib/Immediate/Immediate.cpp#L228) mode to 
+   the compiler, which runs your code in-place via a JIT, rather than compiling it)
+5. **Scriptable extensions** (think about launching "smoke tests" in build pipelines, early, under a JIT while the 
+   project is still building so that you could detect and terminate broken builds earlier)
```

Comment:
I missed this earlier, but for scriptable extensions I think the idea is that you can run a script (under the JIT) against some API surface you expose from your app.

The classic way to extend a precompiled program dynamically is with plugins that are compiled to a shared library and then dlopened. Alternatively people add bindings for some scripting language (e.g. python) and let plugin writers use that, but then you need to manage the foreign function interface between your program's native language / ABI and the scripting language. The nice thing about JIT linking is that you could use it as a basis for writing scripts in the native language of the program.

Reply:
As per offline discussion changed this to -
>  **Scriptable extensions** (think about running JIT’d code in the context of some existing app, allowing the app to be extended by JIT’d code rather than precompiled plugins)